pyvista/core/filters/image_data.py

```diff
@@ -950,6 +951,268 @@ def contour_labeled(
         _vtk.vtkLogger.SetStderrVerbosity(verbosity)
         return cast(pyvista.PolyData, wrap(alg.GetOutput()))
 
+    def point_voxels_to_cell_voxels(self, copy: bool = True):
+        """Convert point voxel data to cell voxel data.
+
+        Convert voxel data represented as points in a uniform grid into voxel cells
+        in a uniform grid. The conversion is performed such that the input points have
+        the same world coordinates as the centers of the converted voxel cells.
+
+        All point data at the input (if any) is passed through unmodified and associated
+        with cell data at the output. The data arrays are copied by default, but this
+        can be disabled to allow both the input (point voxels) and output (cell voxels)
+        to refer to the same data array(s) in memory. Any cell data at the input is
+        ignored and is not used.
+
+        The dimensions of the returned image are all increased by one relative to the
+        input dimensions. The number of cells at the output equals the number of points
+        at the input.
+
+        Since many filters operate on point data exclusively or are inherently cell-based,
+        this conversion enables the same data to be used with either kind of filter while
+        ensuring the input data to those filters has the appropriate representation of
+        the voxels.
+
+        .. versionadded:: 0.44.0
+
+        .. note::
+            This function can be used with any :class:`pyvista.ImageData`, not just
+            3D image data. For example, it can also be used to convert 2D point pixels
+            to cell pixels.
+
+        See Also
+        --------
+        cell_voxels_to_point_voxels
+        :meth:`~pyvista.DataSetFilters.cell_data_to_point_data`
+        :meth:`~pyvista.DataSetFilters.point_data_to_cell_data`
+
+        Parameters
+        ----------
+        copy : bool, default: True
+            Copy the input point data before associating it with the output cell data.
+            If ``False``, the input and output will both refer to the same data arrays.
+
+        Returns
+        -------
+        pyvista.ImageData
+            Image with voxels represented as cells.
+
+        Examples
+        --------
+        Create image data with eight points representing eight point voxels.
+
+        >>> import pyvista as pv
+        >>> voxel_points = pv.ImageData(dimensions=(2, 2, 2))
+        >>> voxel_points.point_data['Data'] = range(8)
+        >>> voxel_points.dimensions
+        (2, 2, 2)
+
+        If we plot the image, it is visually represented as a single voxel **cell** with
+        eight points.
+
+        >>> voxel_points.plot(show_edges=True)
+
+        However, this does not visually show the correct representation when point data
+        is used to represent the centers of voxels. In this case we can convert the
+        point data to cell data to create a cell-based representation of the voxels.
```

Comment:
The actual point or cell data is not modified at all. It is passed through without modification. 

The conversion is completely lossless in that regard. It is possible to convert the point voxels to cell voxels and back any number of times and the data will remain the same. 

What this filter does is simply change how voxels are represented in terms of the actual ImageData container itself. 

Basically, the filter simply modifies the _container_ (I.e the ImageData itself, not the point or cell data), to change what the "definition" of a voxel is.

Many image libraries define an image as follows, for example:

![image](https://github.com/pyvista/pyvista/assets/89109579/1c36ddb9-c2bd-406a-bd41-117b75396d43)

In this case, the image is defined as an array of pixel cells, but the point data of the image is the _center_ of the pixel (or voxel). So, in this case, a pixel or voxel is fundamentally "a point in the center of a pixel cell". Generally, when working with an array of point data, the points themselves represent the center points of the pixels or voxels. 


This contrasts with how VTK fundamentally defines ImageData, where the points are the "corners" of cells, not the cell centers. So, if I have a single point voxel, what I really have is a point in the center of vtk Voxel Cell. So in order to correctly represent a single voxel point , I actually need a single vtk Voxel Cell. The confusing part is that this now means that I need 8 points (of the Voxel Cell) to represent the single point voxel that I actually have (which is at the center of the cell).

Reply:
If I have scalar data for a single voxel (an abstract voxel, not a vtk voxel cell), there are two distinct representations I can use. Both are valid representations but have different uses depending on the filter or application.  

1: Represent this as a single vtk point, with a point data array of length one

2: Represent this as a single vtk voxel cell, with a cell data array of length one.

If you'd like I can provide more concrete examples using specific filters to demonstrate what I mean and show why the filters from this PR are useful.

Reply:
For additional context about the notion of pixels as cells vs pixels as points, and the confusion surrounding it in VTK, see the vtk discourse discussion about it here:

https://discourse.vtk.org/t/proposal-to-add-orientation-to-vtkimagedata-feedback-wanted/120/44

Only the linked comment, and the replies after it are relevant (the rest of the discussion is about image orientation)

Reply:
This quote from Ken Martin summarizes what this PR does very nicely. 

"People tend to expect 100 pixels at 1mm per pixel to create an image of 100mm, but in VTK 100 point samples on an axis equals 99 cells on an axis. And when you draw those cells they look like 99 pixels when people expect 100. I think moving to cell data for pixels would get us to a more common representation. The image would have 100 pixels of cell data, 100 cells, and 101 points defining the boundary of those cells (on an axis)."

`point_voxels_to_cell_voxels` will make it so that 100 point samples (with 99 cells) becomes 100 cells with 101 points. 

`cell_voxels_to_point_voxels` does the reverse.

In either case, the data itself is the same, it's just a matter of representation. And this difference in representation will impact how the filter processes the data. If the wrong representation is used, this can give incorrect or unexpected results with some filters.

Reply:
Thank you. That was a very clear explanation. I will merge once I have your thoughts on the method names.

Reply:
> Great, I will rename the filters and push the updates soon, within the next day or so